File: backend/src/services/report-service.ts
import { getStoredInterviewSessionsBySessionId } from '../storage/interview-session-store.js'
import {
  getStoredInterviewReportBySessionId,
  getStoredInterviewReports,
  getStoredInterviewReportsByOwner,
  upsertStoredInterviewReport
} from '../storage/interview-report-store.js'
import type {
  GenerateInterviewReportRequest,
  GenerateInterviewReportResponse,
  InterviewReportListItem,
  StoredInterviewReportSummary
} from '../types/report.js'
import { buildReportLlmMessages } from '../utils/build-report-llm-messages.js'
import { completeRemoteLlmJson, isRemoteLlmConfigured } from '../utils/complete-remote-llm-json.js'
import { parseReportLlmJson } from '../utils/parse-report-llm-json.js'

const topicLabelMap: Record<string, string> = {
  vue3: 'Vue 3',
  typescript: 'TypeScript',
  engineering: '工程化',
  browser: '浏览器',
  performance: '性能优化',
  scenario: '场景题'
}

const normalizeSnippet = (content: string, maxLength = 88) => {
  const normalized = content.replace(/\s+/g, ' ').trim()
  if (!normalized) return ''
  return normalized.length > maxLength ? `${ normalized.slice(0, maxLength) }...` : normalized
}

const resolvePracticeQuestionType = (weakness: string) => {
  if (/追问|场景|项目|案例|沟通|表达/.test(weakness)) return 'scenario'
  if (/代码|实现|语法|类型|泛型|响应式|性能/.test(weakness)) return 'code'
  return 'concept'
}

const resolvePracticeDifficulty = (answerLength: number, weaknessCount: number) => {
  if (answerLength < 80) return 'easy'
  if (weaknessCount >= 3) return 'hard'
  return 'medium'
}

const resolvePracticeFocusAreas = (
  latestAnswer: string,
  latestFeedback: string,
  primaryWeakness: string
) => {
  const content = `${ primaryWeakness } ${ latestFeedback } ${ latestAnswer }`
  const focusAreas: string[] = []

  if (latestAnswer.trim().length < 90 || /结构|拆分|分段|结论|表达|没讲清|不够完整/.test(content)) {
    focusAreas.push('structure')
  }
  if (/场景|案例|项目|经历|细节|上下文|取舍|过程/.test(content)) {
    focusAreas.push('case_detail')
  }
  if (/结果|指标|收益|数据|效果|量化|验证/.test(content)) {
    focusAreas.push('result_metric')
  }
  if (/原理|为什么|设计|响应式|源码|底层|机制|追问/.test(content)) {
    focusAreas.push('principle_depth')
  }

  return focusAreas.length ? focusAreas : ['structure']
}

const practiceZoneByTopic: Record<string, string> = {
  vue3: 'vue',
  typescript: 'typescript',
  engineering: 'engineering',
  browser: 'javascript',
  performance: 'performance',
  scenario: 'engineering'
}

const buildReportListItem = (report: StoredInterviewReportSummary): InterviewReportListItem => ({
  id: report.id,
  sessionId: report.sessionId,
  threadId: report.threadId,
  topic: report.topic,
  questionTitle: report.questionTitle,
  summaryHeadline: report.summaryHeadline,
  answeredCount: report.answeredCount,
  totalCount: report.totalCount,
  weaknessTags: report.weaknessTags,
  createdAt: report.createdAt,
  updatedAt: report.updatedAt
})

const buildLocalReport = (
  payload: GenerateInterviewReportRequest,
  existing: StoredInterviewReportSummary | null,
  sessions: ReturnType<typeof getStoredInterviewSessionsBySessionId>,
  now: string
): StoredInterviewReportSummary => {
  const allMessages = sessions.flatMap(session => session.messages.map(message => ({
    ...message,
    threadId: session.threadId,
    questionTitle: session.questionTitle
  })))

  const userMessages = allMessages.filter(item => item.role === 'user')
  const feedbackMessages = allMessages.filter(item => item.role === 'assistant')
  const latestAnswer = userMessages[userMessages.length - 1]?.content || ''
  const latestFeedback = feedbackMessages[feedbackMessages.length - 1]?.content || ''
  const answerLength = latestAnswer.trim().length

  const topic = payload.topic?.trim() || sessions[0]?.topic || 'vue3'
  const topicLabel = topicLabelMap[topic] || topic
  const source = payload.source?.trim() || 'mock-interview-space'
  const sourceDocumentName = payload.sourceDocumentName?.trim() || ''
  const sourceLabel = sourceDocumentName || '当前训练上下文'
  const weaknessTags = payload.weaknessTags?.length
    ? [...payload.weaknessTags]
    : []
  const primaryWeakness = payload.primaryWeakness?.trim()
    || weaknessTags[0]
    || '当前还没有形成稳定弱项'

  const answeredCount = typeof payload.answeredCount === 'number'
    ? payload.answeredCount
    : sessions.filter(session => session.messages.some(item => item.role === 'user')).length

  const totalCount = typeof payload.totalCount === 'number' && payload.totalCount > 0
    ? payload.totalCount
    : Math.max(sessions.length, answeredCount, 1)

  const answeredSummary = `${ answeredCount } / ${ totalCount }`
  const questionTitle = sessions[sessions.length - 1]?.questionTitle
    || sessions[0]?.questionTitle
    || topicLabel

  const answerSnapshot = sessions.map((session) => {
    const answerMessage = [...session.messages].reverse().find(item => item.role === 'user')
    const answerText = normalizeSnippet(answerMessage?.content || '', 88)
    return `${ session.questionTitle }: ${ answerText || '未作答' }`
  })

  const summaryHeadline = `围绕 ${ sourceLabel } 的 ${ topicLabel } 训练已形成阶段总结`
  const summaryBody = payload.summaryBody?.trim() || [
    `本轮围绕“${ questionTitle }”共展开 ${ Math.max(sessions.length, totalCount) } 题，当前已完成 ${ answeredSummary }。`,
    latestAnswer
      ? `最近一次作答摘录为“${ normalizeSnippet(latestAnswer, 90) }”，当前最明显的短板是“${ primaryWeakness }”。`
      : `本轮已经浏览完题目，但还存在未作答内容，当前最明显的短板先记为“${ primaryWeakness }”。`
  ].join('')

  const suggestedFocus = payload.suggestedFocus?.length
    ? [...payload.suggestedFocus]
    : [
        answerLength < 80 ? '下一轮回答尽量按“结论 -> 拆分 -> 结果”三段式展开，避免答案过短。' : '',
        latestFeedback ? `优先处理最近一次反馈暴露的问题：${ normalizeSnippet(latestFeedback, 72) }` : '',
        sourceDocumentName ? `下一轮继续围绕《${ sourceDocumentName }》补练，把资料上下文真正转成可表达内容。` : '',
        answeredCount < totalCount ? '当前轮次还有未作答题目，这份报告会先按已展开题目生成阶段性结果。' : ''
      ].filter(Boolean)

  const weaknessFocusAreas = payload.weaknessFocusAreas?.length
    ? [...payload.weaknessFocusAreas]
    : resolvePracticeFocusAreas(latestAnswer, latestFeedback, primaryWeakness)
  const practiceQuestionType = resolvePracticeQuestionType(primaryWeakness)
  const practiceDifficulty = resolvePracticeDifficulty(answerLength, weaknessTags.length)

  return {
    id: `report-${ payload.sessionId }`,
    sessionId: payload.sessionId,
    modelId: payload.modelId?.trim() || existing?.modelId,
    owner: sessions[sessions.length - 1]?.owner || sessions[0]?.owner,
    threadId: sessions[sessions.length - 1]?.threadId,
    topic,
    source,
    sourceDocumentId: payload.sourceDocumentId?.trim() || undefined,
    sourceDocumentName: sourceDocumentName || undefined,
    sourceDocumentExcerpt: payload.sourceDocumentExcerpt?.trim() || undefined,
    questionTitle,
    summaryHeadline,
    summaryBody,
    weaknessTags,
    primaryWeakness,
    weaknessFocusAreas,
    answeredCount,
    totalCount,
    answerSnapshot: [
      ...answerSnapshot,
      latestFeedback ? `系统反馈: ${ normalizeSnippet(latestFeedback, 72) }` : ''
    ].filter(Boolean),
    questionReviews: payload.questionReviews?.length ? payload.questionReviews.map(item => ({ ...item })) : undefined,
    suggestedFocus,
    practicePlan: {
      weaknessTag: primaryWeakness,
      questionType: practiceQuestionType,
      difficulty: practiceDifficulty,
      zone: practiceZoneByTopic[topic] || 'vue'
    },
    createdAt: existing?.createdAt || now,
    updatedAt: now
  }
}

const tryEnhanceReportWithRemoteLlm = async (
  payload: GenerateInterviewReportRequest,
  localReport: StoredInterviewReportSummary
) => {
  if (!isRemoteLlmConfigured(payload.modelId)) return localReport

  const messages = buildReportLlmMessages(payload, {
    topicLabel: topicLabelMap[localReport.topic] || localReport.topic,
    sourceLabel: localReport.sourceDocumentName || '当前训练上下文',
    questionTitle: localReport.questionTitle || (topicLabelMap[localReport.topic] || localReport.topic),
    answeredCount: localReport.answeredCount,
    totalCount: localReport.totalCount,
    answerSnapshot: localReport.answerSnapshot || [],
    latestAnswer: payload.questionReviews?.[payload.questionReviews.length - 1]?.userAnswer || '',
    latestFeedback: payload.questionReviews?.[payload.questionReviews.length - 1]?.aiFeedback || '',
    primaryWeakness: localReport.primaryWeakness || ''
  })

  const raw = await completeRemoteLlmJson(messages, {
    modelId: payload.modelId
  })

  const enhanced = parseReportLlmJson(raw, {
    summaryHeadline: localReport.summaryHeadline,
    summaryBody: localReport.summaryBody,
    primaryWeakness: localReport.primaryWeakness,
    weaknessTags: localReport.weaknessTags,
    weaknessFocusAreas: localReport.weaknessFocusAreas,
    suggestedFocus: localReport.suggestedFocus,
    practicePlan: localReport.practicePlan
  })

  return {
    ...localReport,
    summaryHeadline: enhanced.summaryHeadline,
    summaryBody: enhanced.summaryBody,
    primaryWeakness: enhanced.primaryWeakness,
    weaknessTags: enhanced.weaknessTags,
    weaknessFocusAreas: enhanced.weaknessFocusAreas,
    suggestedFocus: enhanced.suggestedFocus,
    practicePlan: enhanced.practicePlan
  } satisfies StoredInterviewReportSummary
}

export class ReportService {
  listReports(owner?: string): InterviewReportListItem[] {
    const reports = owner
      ? getStoredInterviewReportsByOwner(owner)
      : getStoredInterviewReports().filter(report => !report.owner)
    return reports.map(buildReportListItem)
  }

  getReportBySessionId(sessionId: string, owner?: string) {
    const report = getStoredInterviewReportBySessionId(sessionId)
    if (!report) return null
    if (owner && report.owner !== owner) return null
    if (!owner && report.owner) return null
    return report
  }

  async generateReport(payload: GenerateInterviewReportRequest): Promise<GenerateInterviewReportResponse> {
    const sessionId = payload.sessionId.trim()
    if (!sessionId) {
      throw new Error('sessionId is required.')
    }

    const existing = getStoredInterviewReportBySessionId(sessionId)
    const sessions = getStoredInterviewSessionsBySessionId(sessionId)
    const now = new Date().toISOString()
    const normalizedPayload: GenerateInterviewReportRequest = {
      ...payload,
      sessionId
    }

    const localReport = buildLocalReport(normalizedPayload, existing, sessions, now)

    let finalReport = localReport
    try {
      finalReport = await tryEnhanceReportWithRemoteLlm(normalizedPayload, localReport)
    } catch (error) {
      console.warn('[report-service] remote report generation fallback to local summary:', error)
    }

    upsertStoredInterviewReport(finalReport)

    return {
      report: finalReport,
      created: !existing
    }
  }
}

File: backend/cpp/src/services/ReportService.cpp
#include "services/ReportService.hpp"

#include <chrono>
#include <ctime>
#include <iomanip>
#include <iostream>
#include <sstream>

namespace {
std::string currentTimestampUtc() {
    const auto now = std::chrono::system_clock::now();
    const auto tt = std::chrono::system_clock::to_time_t(now);
    std::tm tm{};
#ifdef _WIN32
    gmtime_s(&tm, &tt);
#else
    gmtime_r(&tt, &tm);
#endif

    std::ostringstream oss;
    oss << std::put_time(&tm, "%Y-%m-%dT%H:%M:%SZ");
    return oss.str();
}

void applyRequestMetadata(InterviewReportEntity& report, const GenerateReportRequest& request) {
    report.modelId = request.modelId;
    report.sourceDocumentExcerpt = request.sourceDocumentExcerpt;
    report.questionReviews = request.questionReviews;
}
}

ReportService::ReportService(ISessionRepository& sessionRepo,
                  IReportRepository& reportRepo,
                  IReportAiClient& aiClient)
                  :m_sessionRepo(sessionRepo),
                  m_reportRepo(reportRepo),
                  m_aiClient(aiClient)
{}

GenerateReportResult ReportService::generateReport(const GenerateReportRequest& request) {
    // 1. 校验 sessionId
    if (request.sessionId.empty()) {
        throw std::runtime_error("sessionId is required");
    }

    // 2. 先查 session
    auto sessions = m_sessionRepo.listSessionsBySessionId(request.sessionId);
    if (sessions.empty()) {
        throw std::runtime_error("Session not found");
    }

    // 3. 取最新一条作为报告上下文
    const auto& session = sessions.back();

    // 4. 判断是否已有报告
    auto existing = m_reportRepo.getReportBySessionId(request.sessionId);

    // 5. 先拼 prompt
    const auto systemPrompt = m_promptBuilder.buildSystemPrompt();
    const auto userPrompt = m_promptBuilder.buildUserPrompt(session, request);

    // 6. 调 AI
    std::string rawJson;
    bool aiSucceeded = true;
    std::string parseError;
    std::string fallbackReason;
    try {
        rawJson = m_aiClient.generateJson(systemPrompt, userPrompt);
    } catch (const std::exception& e) {
        aiSucceeded = false;
        fallbackReason = e.what();
        std::cerr << "[ReportService] AI report generation failed: " << fallbackReason << std::endl;
    }

    // 7. 解析 AI 结果，失败则 fallback
    InterviewReportEntity report;
    if (aiSucceeded) {
        auto parsed = parseAiResultToEntity(rawJson, session, request, parseError);
        if (parsed.has_value()) {
            report = std::move(*parsed);
        } else {
            aiSucceeded = false;
            fallbackReason = "Model output is not valid report JSON: " + parseError;
            std::cerr << "[ReportService] AI report parse failed: " << fallbackReason << std::endl;
            report = buildFallbackReport(session, request);
        }
    } else {
        report = buildFallbackReport(session, request);
    }
    applyRequestMetadata(report, request);

    // 8. 如果已有报告，保留原 id / createdAt
    const auto now = currentTimestampUtc();
    if (existing.has_value()) {
        report.id = existing->id;
        report.createdAt = existing->createdAt;
        report.updatedAt = now;
    } else {
        report.createdAt = now;
        report.updatedAt = now;
    }

    // 9. 落库
    m_reportRepo.upsertReport(report);

    GenerateReportResult result;
    result.report = std::move(report);
    result.created = !existing.has_value();
    result.aiUsed = aiSucceeded;
    if (!aiSucceeded && !fallbackReason.empty()) {
        result.fallbackReason = fallbackReason;
    }
    return result;
}

std::optional<InterviewReportEntity> ReportService::parseAiResultToEntity(
    const std::string& rawJson,
    const InterviewSessionDetail& session,
    const GenerateReportRequest& request,
    std::string& errorMessage)
{
    try {
        auto j = nlohmann::json::parse(rawJson);

        InterviewReportEntity report;
        report.id = "report-" + session.sessionId;
        report.sessionId = session.sessionId;
        report.threadId = session.threadId;
        report.topic = request.topic.value_or(session.topic);
        report.source = request.source.value_or("mock-interview-space");
        report.sourceDocumentId = request.sourceDocumentId;
        report.sourceDocumentName = request.sourceDocumentName;
        report.modelId = request.modelId;
        report.sourceDocumentExcerpt = request.sourceDocumentExcerpt;
        report.questionReviews = request.questionReviews;
        report.questionTitle = session.questionTitle;

        report.summaryHeadline = j.value("summaryHeadline", "");
        report.summaryBody = j.value("summaryBody", "");
        report.weaknessTags = j.value("weaknessTags", std::vector<std::string>{});
        if (j.contains("primaryWeakness") && j["primaryWeakness"].is_string()) {
            report.primaryWeakness = j["primaryWeakness"].get<std::string>();
        }

        if (j.contains("answeredCount") && j["answeredCount"].is_number_integer()) {
            report.answeredCount = j["answeredCount"].get<int>();
        } else {
            report.answeredCount = request.answeredCount.value_or(0);
        }

        if (j.contains("totalCount") && j["totalCount"].is_number_integer()) {
            report.totalCount = j["totalCount"].get<int>();
        } else {
            report.totalCount = request.totalCount.value_or(0);
        }

        if (j.contains("suggestedFocus") && j["suggestedFocus"].is_array()) {
            report.suggestedFocus = j["suggestedFocus"].get<std::vector<std::string>>();
        }

        if (j.contains("practicePlan") && j["practicePlan"].is_object()) {
            PracticePlanSnapshot plan;
            const auto& p = j["practicePlan"];
            plan.weaknessTag = p.value("weaknessTag", report.primaryWeakness.value_or("当前弱项"));
            plan.questionType = p.value("questionType", "concept");
            plan.difficulty = p.value("difficulty", "medium");
            plan.zone = p.value("zone", "vue");
            report.practicePlan = plan;
        }

        report.summaryHeadline = report.summaryHeadline.empty()
            ? "阶段性复盘总结"
            : report.summaryHeadline;

        return report;
    } catch (const std::exception& e) {
        errorMessage = e.what();
        return std::nullopt;
    }
}

InterviewReportEntity ReportService::buildFallbackReport(
    const InterviewSessionDetail& session,
    const GenerateReportRequest& request) const
{
    InterviewReportEntity report;
    report.id = "report-" + session.sessionId;
    report.sessionId = session.sessionId;
    report.threadId = session.threadId;
    report.topic = request.topic.value_or(session.topic);
    report.source = request.source.value_or("mock-interview-space");
    report.sourceDocumentId = request.sourceDocumentId;
    report.sourceDocumentName = request.sourceDocumentName;
    report.modelId = request.modelId;
    report.sourceDocumentExcerpt = request.sourceDocumentExcerpt;
    report.questionReviews = request.questionReviews;
    report.questionTitle = session.questionTitle;

    report.answeredCount = request.answeredCount.value_or(session.messageCount / 2);
    report.totalCount = request.totalCount.value_or(session.messageCount);
    report.weaknessTags = request.weaknessTags.value_or(std::vector<std::string>{});
    if (request.primaryWeakness.has_value()) {
        report.primaryWeakness = request.primaryWeakness;
    } else if (!report.weaknessTags.empty()) {
        report.primaryWeakness = report.weaknessTags.front();
    }

    report.summaryHeadline = "本轮训练已完成基础复盘";
    report.summaryBody = request.summaryBody.value_or("本轮对话已沉淀为结构化报告，建议继续针对弱项训练。");

    if (request.suggestedFocus.has_value()) {
        report.suggestedFocus = request.suggestedFocus;
    }

    if (request.weaknessFocusAreas.has_value()) {
        report.weaknessFocusAreas = request.weaknessFocusAreas;
    }

    if (request.sourceDocumentExcerpt.has_value()) {
        report.answerSnapshot = std::vector<std::string>{ *request.sourceDocumentExcerpt };
    }

    report.practicePlan = PracticePlanSnapshot{
        report.primaryWeakness.value_or("当前弱项"),
        "concept",
        "medium",
        "vue"
    };

    return report;
}

std::vector<InterviewReportSummary> ReportService::listReports() {
    return m_reportRepo.listReports();
}

std::optional<InterviewReportEntity> ReportService::getReportBySessionId(const std::string& sessionId) {
    if (sessionId.empty()) {
        return std::nullopt;
    }
    return m_reportRepo.getReportBySessionId(sessionId);
}

void ReportService::clearAllReports() {
    m_reportRepo.clearAll();
}
